sanitycheck: device handler, allow running tests on real hw

nashif opened the pull request:
This will allow us to run sanitycheck on real devices and get reporting
out of it the same way we do that with Qemu.

To use this, run sanitycheck with the following new options:

 ```
scripts/sanitycheck --device-testing --device-serial /dev/ttyACM0 -p frdm_k64f  -T tests/crypto/ -T tests/kernel --detailed-report frdm_k64f.xml
junit2html frdm_k64f.xml
```

--device-serial denotes the serial device the board is connected to.
This needs to be accessible by the user running sanitycheck. You can
run this on one board only at a time, the board is specified using the
--platform option. (-p)

This was tested with only a few boards, some board will not work
because how they reset the serial device during flashing.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: parse test results and create detailed report

Parse the test results and create a test report with more granular
results that can be imported to into test management/reporting system.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: device handler, allow running tests on real hw

This will allow us to run sanitycheck on real devices and get reporting
out of it the same way we do that with Qemu.

To use this, run sanitycheck with the following new options:

 scripts/sanitycheck --device-testing --device-serial /dev/ttyACM0 -p
 frdm_k64f  -T tests/crypto/

--device-serial denotes the serial device the board is connected to.
This needs to be accessible by the user running sanitycheck. You can
run this on one board only at a time, the board is specified using the
--platform option.

This was tested with only a few boards, some board will not work
because how they reset the serial device during flashing.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

carlescufi commented:
cc @myrdyr @grho @TomaszLukanko

galak reviewed (commented):
Do we need to update scripts/requirements.txt for pyserial?

nashif commented:
> Do we need to update scripts/requirements.txt for pyserial?

yes, need to update a few other things like documentation and CI image as well, all will be in the final PR

galak commented:
Is it feasible to make this also work with a prebuilt set of binaries.  Something like doing a sanitycheck build first than doing a sanitycheck run.

nashif commented:
> Is it feasible to make this also work with a prebuilt set of binaries. Something like doing a sanitycheck build first than doing a sanitycheck run.

will be very complex and difficult, the feature depends heavily on cmake and the artefacts it generates to be able to call 'make flash', so just pointing to a binary will not work and probably need to be addressed with another tool. Also the reporting and everything else we get with sanitycheck will be difficult to reproduce and get right.

MaureenHelm commented on scripts/sanitycheck:
```diff
@@ -316,6 +317,86 @@ class Handler:
+        self.timeout = instance.test.timeout
+        self.sourcedir = instance.test.code_location
+        self.outdir = instance.outdir
+        self.run_log = os.path.join(self.outdir, "run.log")
```
I don't see this being used anywhere. How is it different from handler_log?

nashif replied:
right, was a copy from another handler, need to refactor the handler class better and stop redeclaring this in subclasses. I have this refactoring on the TODO.

MaureenHelm commented on scripts/sanitycheck:
```diff
@@ -316,6 +317,86 @@ class Handler:
+            ser.close()
+
+        if harness.state:
+            self.set_state(harness.state, {})
```
Is this parsing the serial output for pass/fail?

nashif replied:
yes, in scripts/sanity_chk/harness.py

nashif committed:
sanitycheck: cleanup handler class

init class in one place, no need to duplicate all class members in every
subclass.
run_log is not needed in the handler class.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: simplify argument passing and use global options

We are passing global arguments from one level to the next when those
variables are available globally. Reduce the arguments and remove unused
arguments as well.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: remove dot after PASS|FAIL

That dot does not belong here, just stands in the way when parsing.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: support skipped tests, enhance device handler

- Some tests start with test_, some do not, so make sure we parse both.
- Parse skipped tests
- Improve handling of test case identifier
- Handle Exceptions in device handler

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

carlescufi commented:
@myrdyr: Can you take a look and copy Grzegorz as well?

nashif committed:
sanitycheck: do not count duplicate tests

Some tests are being reused to reset or re-init state, ignore them in
accounting.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
ci: remove pyserial install

This is now installed in the docker image.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: we need pyserial for sanitycheck

Needed for testing on real hardware and connecting to the serial console
to capture output.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

carlescufi reviewed (approved):
I have tested this on Nordic hardware with decent results.

carlescufi commented on scripts/sanitycheck:
```diff
@@ -2043,6 +2138,79 @@ class TestSuite:
+                        eleTestcase,
+                        'skipped')
+
+        result = ET.tostring(eleTestsuites)
```
There's no CR/LF in the saved file. Is this intentional?

nashif replied:
just run it through xmllint --format :)
Those files are not intended to be read by humans.
